app/services/targets/create_service.rb

```diff
@@ -11,13 +11,32 @@ def initialize(current_user, params)
     def call
       raise MaxAllowTargetsError unless targets.count < MAX_ALLOWED_TARGETS
 
-      targets.create!(params)
+      target = targets.create!(params)
+      find_matches(target)
+      target
     end
 
     private
 
     def targets
       @targets ||= current_user.targets
     end
+
+    def find_matches(target)
+      compatible_target = Target.within(target.radius, origin: [target.latitude, target.longitude])
+                                .where(topic_id: target.topic_id)
+                                .where.not(user_id: target.user_id)
+                                 &.first
+      create_match(target, compatible_target) if compatible_target.present?
+    end
+
+    def create_match(target, compatible_target)
+      target_matcher = TargetMatcher.create!(topic_id: target.topic_id)
+
+      TargetMatcherInstance.create!(user: target.user, target_matcher: target_matcher,
+                                    target: target)
+      TargetMatcherInstance.create!(user: compatible_target.user, target_matcher: target_matcher,
+                                    target: compatible_target.target)
```

Comment:
```suggestion
      [target, compatible_target].each do |t|
          TargetMatcherInstance.create!(user: t.user, target: t, target_matcher: target_matcher)
      end
```
Here you could do something like that. In this case I create a list with the 2 target we are interested on create a TargetMatcherInstance, and iterate over that list, using a generic way to create both Instance without repeat code.